http/oidc/src/main/java/org/wildfly/security/http/oidc/OidcRequestAuthenticator.java

```diff
@@ -416,4 +484,112 @@ private static boolean hasScope(String scopeParam, String targetScope) {
+                    deployment.getClientKeyStorePassword(), deployment.getClientKeyPassword(),
+                    deployment.getClientKeyAlias(), deployment.getClientKeystoreType());
+        }
+    }
+
+    public JsonWebSignature signRequest(JwtClaims jwtClaims) throws Exception {
+        JsonWebSignature jsonWebSignature = new JsonWebSignature();
+        jsonWebSignature.setPayload(jwtClaims.toJson());
+
+        if (deployment.getRequestSignatureAlgorithm().contains(NONE)) { //unsigned
+            jsonWebSignature.setAlgorithmConstraints(AlgorithmConstraints.NO_CONSTRAINTS);
+            jsonWebSignature.setAlgorithmHeaderValue(NONE);
+        } else if (deployment.getRequestSignatureAlgorithm().contains(HS_256)
+                || deployment.getRequestSignatureAlgorithm().contains(HS_384)
+                || deployment.getRequestSignatureAlgorithm().contains(HS_512)) { //signed with symmetric key
+            jsonWebSignature.setAlgorithmHeaderValue(deployment.getRequestSignatureAlgorithm());
+            Key key = new HmacKey(deployment.getResourceCredentials().get("secret").toString().getBytes(StandardCharsets.UTF_8));   //the client secret is a shared secret between the server and the client
+            jsonWebSignature.setDoKeyValidation(false);     //skips validation so that size of the secret does not matter
+            jsonWebSignature.setKey(key);
+        } else { //signed with asymmetric key
+            KeyPair keyPair = getkeyPair();
+            jsonWebSignature.setKey(keyPair.getPrivate());
+            jsonWebSignature.setAlgorithmConstraints(new AlgorithmConstraints(PERMIT, deployment.getRequestSignatureAlgorithm()));
+            jsonWebSignature.setAlgorithmHeaderValue(deployment.getRequestSignatureAlgorithm());
+        }
+        jsonWebSignature.sign();
+        return jsonWebSignature;
+    }
+
+    private JsonWebEncryption encryptRequest(JsonWebSignature signedRequest) throws JoseException {
+        JsonWebEncryption jsonEncryption = new JsonWebEncryption();
+        jsonEncryption.setPayload(signedRequest.getCompactSerialization());
+        jsonEncryption.setAlgorithmConstraints(new AlgorithmConstraints(AlgorithmConstraints.ConstraintType.PERMIT, deployment.getRequestEncryptAlgorithm(), deployment.getRequestEncryptEncValue()));
+        jsonEncryption.setAlgorithmHeaderValue(deployment.getRequestEncryptAlgorithm());
+        jsonEncryption.setEncryptionMethodHeaderParameter(deployment.getRequestEncryptEncValue());
+        List<JsonWebKey> jwkList = deployment.getrealmKeySet().getJsonWebKeys();
+        jsonEncryption.setDoKeyValidation(false);
+        for (JsonWebKey jwk : jwkList) {
+            if (deployment.getRequestEncryptAlgorithm().contains(RSA_OAEP) || deployment.getRequestEncryptAlgorithm().contains(RSA_OAEP_256) || deployment.getRequestEncryptAlgorithm().contains(RSA1_5)) {
+                if (jwk.getUse().equals("enc")) {  //JWT's for keycloak are to be encrypted with realm public keys
```

Comment:
Do other providers also use "enc" to specify the keys to be used for encryption?

Reply:
It seems like there are only two values(sig and enc) according to the JWK specs: https://www.ietf.org/rfc/rfc7517.txt (please see section 4.2)
So I am assuming yes? However, it is optional. So, there may be a case where use is not specified. 
I am not sure what to do in that case, through